extensions/amp-truncate-text/0.1/amp-truncate-text.js

```diff
@@ -0,0 +1,259 @@
+ * - If we had some rough bucket of performance, maybe just fallback
+ *   immediately to gradient / hard cut off.
+ * - Custom fonts can cause truncation to end up being wrong
+ *   when they load
+ *     * Can we just wait to layout if we know a font is loading?
+ *       Since all fonts are statically declared in AMP, this is just a
+ *       one time thing
+ */
+export class AmpTruncateText extends AMP.BaseElement {
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.content_ = null;
+
+    /** @private {?Element} */
+    this.expandSlot_ = null;
+
+    /** @private {?Element} */
+    this.collapseSlot_ = null;
+
+    /** @private {boolean} */
+    this.useShadow_ = false;
+
+    /** @private {?MutationObserver} */
+    this.mutationObserver_ = null;
+
+    /** @private {boolean} */
+    this.truncateRequested_ = false;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.useShadow_ = !!this.element.attachShadow &&
+      isExperimentOn(this.win, 'amp-truncate-text-shadow');
+
+    if (this.useShadow_) {
+      this.buildShadow_();
+    } else {
+      this.build_();
+    }
+
+    if ('MutationObserver' in window) {
+      this.mutationObserver_ = new MutationObserver(() => {
+        this.truncate_();
+      });
+    }
+
+    this.expandSlot_.addEventListener('click', () => this.expand_());
+    this.collapseSlot_.addEventListener('click', () => this.collapse_());
+  }
+
+  /**
+   * Builds the component when not using Shadow DOM.
+   */
+  build_() {
+    const html = htmlFor(this.element);
+    this.content_ = html`
+      <div class="i-amphtml-truncate-content">
+        <span class="i-amphtml-default-slot"></span>
+        <span class="i-amphtml-truncate-expand-slot" name="expand"></span>
+        <span class="i-amphtml-truncate-collapse-slot" name="collapse"></span>
+      </div>
+    `;
+
+    const defaultSlot = this.content_.querySelector('.i-amphtml-default-slot');
+    this.expandSlot_ = this.content_
+        .querySelector('.i-amphtml-truncate-expand-slot');
+    this.collapseSlot_ = this.content_
+        .querySelector('.i-amphtml-truncate-collapse-slot');
+
+    iterateCursor(this.element.querySelectorAll('[slot="expand"]'), el => {
+      this.expandSlot_.appendChild(el);
+    });
+    iterateCursor(this.element.querySelectorAll('[slot="collapse"]'), el => {
+      this.collapseSlot_.appendChild(el);
+    });
+    this.getRealChildNodes().forEach(node => {
+      defaultSlot.appendChild(node);
+    });
+
+    this.element.appendChild(this.content_);
+  }
+
+  /**
+   * Builds the component when using Shadow DOM.
+   */
+  buildShadow_() {
+    // TODO(sparhami) Where is the right place to put this? Runtime? What about
+    // SSR?
+    const sizer = this.element.querySelector('i-amphtml-sizer');
+    if (sizer) {
+      sizer.setAttribute('slot', 'sizer');
+    }
+
+    // TODO(sparhami) Is there a shared place to add logic for creating
```

Comment:
Agree. We need a place for this to use constructable stylesheets and things like `sizer`. May be a good idea to put it in a separate module right in this extension? Would be easier to collect the generic requirements that way.